### PR TITLE
add native param types

### PR DESCRIPTION
**Pull request by orklah**
Following #4116, here's a PR proposing adding native types in parameters.

As before, I used phpstorm to add them and psalm to refine.

### CODE REVIEW

File: src/Psalm/Codebase.php [hunk 1]
```diff
@@ -504,21 +498,15 @@ public function getStatementsForFile($file_path)
...
         return $this->classlike_storage_provider->create($fq_classlike_name);
     }
 
     /**
-     * @param  string $file_path
-     *
      * @return void
      */
-    public function cacheClassLikeStorage(ClassLikeStorage $classlike_storage, $file_path)
+    public function cacheClassLikeStorage(ClassLikeStorage $classlike_storage, string $file_path)
```
Comment (azjezz) [inline src/Psalm/Codebase.php:509]: i think it is safe to add `: void` as well :)

File: src/Psalm/Codebase.php [hunk 2]
```diff
@@ -675,23 +639,16 @@ public function getAllStubbedConstants(): array
...
-     *
-     */
-    public function fileExists($file_path): bool
+    public function fileExists(string $file_path): bool
     {
         return $this->file_provider->fileExists($file_path);
     }
 
     /**
      * Check whether a class/interface exists
```
Comment (azjezz) [inline src/Psalm/Codebase.php:648]: I believe the method name already explains this, so maybe it should be removed?

File: src/Psalm/Codebase.php [hunk 3]
```diff
@@ -704,25 +661,17 @@ public function classOrInterfaceExists(
...
-     */
-    public function classExtendsOrImplements($fq_class_name, $possible_parent): bool
+    public function classExtendsOrImplements(string $fq_class_name, string $possible_parent): bool
     {
         return $this->classlikes->classExtends($fq_class_name, $possible_parent)
             || $this->classlikes->classImplements($fq_class_name, $possible_parent);
     }
 
     /**
      * Determine whether or not a given class exists
```
Comment (azjezz) [inline src/Psalm/Codebase.php:671]: same here

File: src/Psalm/Codebase.php [hunk 4]
```diff
@@ -738,36 +687,24 @@ public function classExists(
...
-     *
      */
-    public function classExtends($fq_class_name, $possible_parent): bool
+    public function classExtends(string $fq_class_name, string $possible_parent): bool
     {
         return $this->classlikes->classExtends($fq_class_name, $possible_parent, true);
     }
 
     /**
      * Check whether a class implements an interface
```
Comment (azjezz) [inline src/Psalm/Codebase.php:699]: same.

File: src/Psalm/Config.php [hunk 5]
```diff
@@ -1172,11 +1158,9 @@ private function loadFileExtensions($extensions)
     }
 
     /**
-     * @param string $path
-     *
      * @return void
      */
-    public function addPluginPath($path)
+    public function addPluginPath(string $path)
```
Comment (azjezz) [inline src/Psalm/Config.php:1163]: `: void`

File: src/Psalm/Config.php [hunk 6]
```diff
@@ -1292,12 +1276,11 @@ private static function requirePath(string $path) : void
     /**
      * @template T
      *
-     * @param  string $path
      * @param  T::class $must_extend
      *
      * @return class-string<T>
      */
-    private function getPluginClassForPath(Codebase $codebase, $path, $must_extend)
+    private function getPluginClassForPath(Codebase $codebase, string $path, $must_extend)
```
Comment (azjezz) [inline src/Psalm/Config.php:1283]: `string` type can be added to `$must_extend` while keeping `T::class` in the docblock to verify that it is a class name.

File: src/Psalm/Config.php [hunk 7]
```diff
@@ -1592,25 +1550,19 @@ public function getReportingLevelForFile($issue_type, $file_path): string
...
     }
 
     /**
-     * @param   string $issue_type
-     * @param   string $method_id
-     *
      * @return  string|null
      */
-    public function getReportingLevelForMethod($issue_type, $method_id)
+    public function getReportingLevelForMethod(string $issue_type, string $method_id)
```
Comment (azjezz) [inline src/Psalm/Config.php:1565]: `: ?string`

**Review by azjezz — APPROVED** [review]
LGTM 👍🏼 just a few comments

File: src/Psalm/Config.php [hunk 8]
```diff
@@ -709,14 +701,11 @@ public static function loadFromXMLFile($file_path, $current_dir): Config
...
      * Creates a new config object from an XML string
-     *
-     * @param  string           $base_dir
-     * @param  string           $file_contents
      * @param  string|null      $current_dir Current working directory, if different to $base_dir
      *
      * @throws ConfigException
      */
-    public static function loadFromXML($base_dir, $file_contents, $current_dir = null): Config
+    public static function loadFromXML(string $base_dir, string $file_contents, $current_dir = null): Config
```
Comment (weirdan) [inline src/Psalm/Config.php:708]: I think this needs to be promoted to `?string $current_dir = null` with corresponding docblock tag dropped. There are many instances of this syntax (old form of nullable params) so it might make sense to do them all at once.

**Comment by orklah** [comment]
@azjezz Thank you for your review. I added the missing param type in config, but I didn't add missing return types. I just made a PR about return types and I prefer to keep things separated. I'll maybe create another PR later for return types that weren't included in last PR (because of complications of edge cases).

@weirdan Thanks too for your review, you're right. I just made a commit with nullable types. I hope I caught them all!

File: src/Psalm/Internal/Analyzer/ClassAnalyzer.php [hunk 9]
```diff
@@ -1763,7 +1753,6 @@ private static function addOrUpdatePropertyType(
     }
 
     /**
-     * @param  Context|null                    $global_context
      * @param  bool                            $is_fake
```
Comment (weirdan) [inline src/Psalm/Internal/Analyzer/ClassAnalyzer.php:1756]:
```suggestion
```

File: src/Psalm/Internal/Provider/StatementsProvider.php [hunk 10]
```diff
@@ -357,20 +355,19 @@ public function resetDiffs()
     }
 
     /**
-     * @param  string  $file_contents
      * @param  string   $file_path
```
Comment (weirdan) [inline src/Psalm/Internal/Provider/StatementsProvider.php:358]:
```suggestion
```

File: src/Psalm/Storage/FunctionLikeParameter.php [hunk 11]
```diff
@@ -94,24 +94,21 @@ class FunctionLikeParameter
     public $expect_variable = false;
 
     /**
-     * @param string        $name
      * @param Type\Union|null    $type
```
Comment (weirdan) [inline src/Psalm/Storage/FunctionLikeParameter.php:97]:
```suggestion
```

File: src/Psalm/Storage/FunctionLikeParameter.php [hunk 12]
```diff
@@ -94,24 +94,21 @@ class FunctionLikeParameter
...
-     * @param CodeLocation|null  $location
      * @param bool       $is_optional
      * @param bool       $is_nullable
      * @param bool       $is_variadic
```
Comment (weirdan) [inline src/Psalm/Storage/FunctionLikeParameter.php:100]:
```suggestion
```

File: src/Psalm/Storage/FunctionLikeParameter.php [hunk 13]
```diff
@@ -94,24 +94,21 @@ class FunctionLikeParameter
...
-     * @param Type\Union|null    $default_type
      */
     public function __construct(
-        $name,
+        string $name,
         bool $by_ref,
         Type\Union $type = null,
```
Comment (weirdan) [inline src/Psalm/Storage/FunctionLikeParameter.php:105]:
```suggestion
        ?Type\Union $type = null,
```

File: src/Psalm/Storage/FunctionLikeParameter.php [hunk 14]
```diff
@@ -94,24 +94,21 @@ class FunctionLikeParameter
...
-        CodeLocation $location = null,
-        CodeLocation $type_location = null,
+        ?CodeLocation $location = null,
+        ?CodeLocation $type_location = null,
         $is_optional = true,
         $is_nullable = false,
         $is_variadic = false,
```
Comment (weirdan) [inline src/Psalm/Storage/FunctionLikeParameter.php:110]:
```suggestion
        bool $is_optional = true,
        bool $is_nullable = false,
        bool $is_variadic = false,
```

**Comment by weirdan** [comment]
> @weirdan Thanks too for your review, you're right. I just made a commit with nullable types. I hope I caught them all! I'll made another commit for redundant types that resulted

Thanks for addressing that. I'm almost done with the review (there are some more nitpicks, but most of them in the form of github suggestions which can be applied with a single click). Sneak peek: 
image

File: src/Psalm/Type/Atomic/CallableTrait.php [hunk 15]
```diff
@@ -40,8 +40,8 @@ trait CallableTrait
      */
     public function __construct(
         $value = 'callable',
```
Comment (weirdan) [inline src/Psalm/Type/Atomic/CallableTrait.php:42]:
```suggestion
        string $value = 'callable',
```
and corresponding `@param` can be dropped

File: tests/fixtures/stubs/DomainException.php [hunk 16]
```diff
@@ -13,7 +13,7 @@ class DomainException extends LogicException
      *
      * @since 5.1.0
      */
-    public function __construct($message = '', $code = 0, Throwable $previous = null)
+    public function __construct($message = '', $code = 0, ?Throwable $previous = null)
```
Comment (weirdan) [inline tests/fixtures/stubs/DomainException.php:16]: I think none of the files in `fixtures` folder should be changed

File: tests/fixtures/stubs/extending_plugin_entrypoint.php [hunk 17]
```diff
@@ -8,7 +8,7 @@
 class ExtendingPluginRegistration implements PluginEntryPointInterface
 {
     /** @return void */
-    public function __invoke(RegistrationInterface $r, SimpleXMLElement $config = null)
+    public function __invoke(RegistrationInterface $r, ?SimpleXMLElement $config = null)
```
Comment (weirdan) [inline tests/fixtures/stubs/extending_plugin_entrypoint.php:11]: Well, this one should

**Review by weirdan — CHANGES_REQUESTED** [review]
Here's some more suggested changes. Some of those may be already outdated, as you pushed more commits during the review.

File: tests/fixtures/stubs/extending_plugin_entrypoint.php (hunk 17 above)
Comment (azjezz) [reply tests/fixtures/stubs/extending_plugin_entrypoint.php:11]: Why this but not `tests/fixtures/stubs/DomainException.php` ? isn't it the same case ( changing the argument typing using the default value )?

**Comment by weirdan** [comment]
>  I'll maybe create another PR later for return types that weren't included in last PR (because of complications of edge cases).

When/if you do it makes sense to address `:void` and nullable return types as part of that. I thinks there's a lot of them still in docblocks, and it's a kind of low-hanging fruit.

File: tests/fixtures/stubs/extending_plugin_entrypoint.php (hunk 17 above)
Comment (weirdan) [reply tests/fixtures/stubs/extending_plugin_entrypoint.php:11]: The signature here is dictated by `PluginEntryPointInterface` that got changed.

**Comment by orklah** [comment]
wow, that's a big review!

I'll take a look at all that soon

**Comment by orklah** [comment]
Ok, I think I've fixed everything.

May I ask how you found those? Only by watching the review page on github? They were mainly params I left alone because either they were in a file I didn't want to touch or phpstorm warned me about some incompatibility (or flat out didn't include a fix for some reason)

**Comment by weirdan** [comment]
> May I ask how you found those? Only by watching the review page on github? 

Yes.

**Comment by orklah** [comment]
> > May I ask how you found those? Only by watching the review page on github?
> 
> Yes.

No wonder that took you some time :) ! Great work!

**Review by weirdan — CHANGES_REQUESTED** [review]
Just one more little fix.

Thanks @orklah, this was a mountainous task.

**Comment by orklah** [comment]
Well, refactoring tools help a lot. Alongside with Psalm being Psalm, when it analyse itself, it ensure it works correcttly and types are coherent at the same time. The massive amount of tests do the rest :)

This wouldn't have been this easy on a lot of codebases (my work codebase to start with!)

**Comment by muglug** [comment]
Thanks! And I agree, having Psalm helps when refactoring Psalm!
